### DISCUSSION TITLE
Unable to run a stub using JIRA Open api yaml

### DISCUSSION THREAD

**Discussion by mahesh-khanna-19855** [specmatic/specmatic#841] [Ask For Help]
I'm in the process of generating a stub for Jira using the Specmatic tool and their OpenAPI. However, I encountered an error when running the stub command.

> java -jar specmatic.jar stub ../jira_openapi.yaml

> Loading ../jira_openapi.yaml Could not load /Users/mahesh/Downloads/jira_openapi.yaml: Specmatic only supports oauth2, bearer, and api key authentication (header, query) security schemes at the moment

Additionally, upon attempting to use the generated stub, I encountered a related exception

`Empty collection can't be reduced java.lang.UnsupportedOperationException: Empty collection can't be reduced. at in.specmatic.stub.HttpStubKt.fakeHttpResponse(HttpStub.kt:1037) at in.specmatic.stub.HttpStubKt.getHttpResponse(HttpStub.kt:648) at in.specmatic.stub.HttpStub.serveStubResponse(HttpStub.kt:312) at in.specmatic.stub.HttpStub.access$serveStubResponse(HttpStub.kt:41) at in.specmatic.stub.HttpStub$environment$1$1$2.invokeSuspend(HttpStub.kt:170) at in.specmatic.stub.HttpStub$environment$1$1$2.invoke(HttpStub.kt) at in.specmatic.stub.HttpStub$environment$1$1$2.invoke(HttpStub.kt) at io.ktor.util.pipeline.SuspendFunctionGun.loop(SuspendFunctionGun.kt:120) at io.ktor.util.pipeline.SuspendFunctionGun.proceed(SuspendFunctionGun.kt:78) at io.ktor.server.engine.BaseApplicationEngineKt$installDefaultTransformationChecker$1.invokeSuspend(BaseApplicationEngine.kt:124) at io.ktor.server.engine.BaseApplicationEngineKt$installDefaultTransformationChecker$1.invoke(BaseApplicationEngine.kt) at io.ktor.server.engine.BaseApplicationEngineKt$installDefaultTransformationChecker$1.invoke(BaseApplicationEngine.kt) at io.ktor.util.pipeline.SuspendFunctionGun.loop(SuspendFunctionGun.kt:120) at io.ktor.util.pipeline.SuspendFunctionGun.proceed(SuspendFunctionGun.kt:78) at io.ktor.util.pipeline.SuspendFunctionGun.execute$ktor_utils(SuspendFunctionGun.kt:98) at io.ktor.util.pipeline.Pipeline.execute(Pipeline.kt:77) at io.ktor.server.engine.DefaultEnginePipelineKt$defaultEnginePipeline$1$invokeSuspend$$inlined$execute$1.invokeSuspend(Pipeline.kt:478) at io.ktor.server.engine.DefaultEnginePipelineKt$defaultEnginePipeline$1$invokeSuspend$$inlined$execute$1.invoke(Pipeline.kt) at io.ktor.server.engine.DefaultEnginePipelineKt$defaultEnginePipeline$1$invokeSuspend$$inlined$execute$1.invoke(Pipeline.kt) at io.ktor.util.debug.ContextUtilsKt.initContextInDebugMode(ContextUtils.kt:17) at io.ktor.server.engine.DefaultEnginePipelineKt$defaultEnginePipeline$1.invokeSuspend(DefaultEnginePipeline.kt:123) at io.ktor.server.engine.DefaultEnginePipelineKt$defaultEnginePipeline$1.invoke(DefaultEnginePipeline.kt) at io.ktor.server.engine.DefaultEnginePipelineKt$defaultEnginePipeline$1.invoke(DefaultEnginePipeline.kt) at io.ktor.util.pipeline.SuspendFunctionGun.loop(SuspendFunctionGun.kt:120) at io.ktor.util.pipeline.SuspendFunctionGun.proceed(SuspendFunctionGun.kt:78) at io.ktor.util.pipeline.SuspendFunctionGun.execute$ktor_utils(SuspendFunctionGun.kt:98) at io.ktor.util.pipeline.Pipeline.execute(Pipeline.kt:77) at io.ktor.server.netty.NettyApplicationCallHandler$handleRequest$1$invokeSuspend$$inlined$execute$1.invokeSuspend(Pipeline.kt:478) at io.ktor.server.netty.NettyApplicationCallHandler$handleRequest$1$invokeSuspend$$inlined$execute$1.invoke(Pipeline.kt) at io.ktor.server.netty.NettyApplicationCallHandler$handleRequest$1$invokeSuspend$$inlined$execute$1.invoke(Pipeline.kt) at io.ktor.util.debug.ContextUtilsKt.initContextInDebugMode(ContextUtils.kt:17) at io.ktor.server.netty.NettyApplicationCallHandler$handleRequest$1.invokeSuspend(NettyApplicationCallHandler.kt:119) at io.ktor.server.netty.NettyApplicationCallHandler$handleRequest$1.invoke(NettyApplicationCallHandler.kt) at io.ktor.server.netty.NettyApplicationCallHandler$handleRequest$1.invoke(NettyApplicationCallHandler.kt) at kotlinx.coroutines.intrinsics.UndispatchedKt.startCoroutineUndispatched(Undispatched.kt:44) at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:112) at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)`

**Comment by harikrishnan83** [comment]
hello @maheshkhanna, thanks for reporting the issue. We are taking a look.

**Comment by harikrishnan83** [comment]
Hello Mahesh,

You may be seeing the error message "Specmatic only supports oauth2, bearer, and api key authentication..." because Jira OpenAPI specifications seem to be using basic auth for some of their endpoints. Basic auth support in Specmatic is currently under development and I will keep you posted on the timelines for the same.

Can you please elaborate "upon attempting to use the generated stub"? It will be helpful if you can share the Specmatic command you ran to start the stub server.

Thanks.

**Reply by mahesh-khanna-19855** [reply]
My application has a feature that is integrated with Jira. To test my application in isolation, I'm attempting to stub it using the Jira Open API. To do this, I'm running the Specmatic JAR using the command below:

`java -jar specmatic.jar stub ../jira_openapi.yaml`

JIRA Open api : https://dac-static.atlassian.com/cloud/jira/platform/swagger.v3.json?_v=1.6733.0-0.1296.0

**Reply by harikrishnan83** [reply]
Thanks for your response. The OpenAPI specification in the link you have shared above confirms the fact that it has basic auth enabled on some of that paths which is why, as I mentioned earlier, you are seeing "Specmatic only supports oauth2, bearer, and api key authentication (header, query) security schemes at the moment". This should be resolved when we add the support for basic auth.

About the exception that you have shared in your initial post with regard to recording the specification, can you please share the steps that you followed. This will help us reproduce the issue at our end.

UPDATE: I am guessing that the exception that you have shared earlier is seen when you hit the stub URL (http://localhost:9000) after you saw the error "Specmatic only supports oauth2, bearer, and api key authentication (header, query) security schemes at the moment". Please correct me if I am wrong. Since no specification was loaded because of the error with basic auth security scheme this may be happening. We are investigating the same.

That said, it will still be helpful if you can share how you recored the API specification.

**Comment by yogeshnikam671** [comment]
Basic Auth is now supported by Specmatic. Here is the [documentation](https://specmatic.in/documentation/authentication.html#basic-authentication) for the same. 
Can you please try with the latest version of Specmatic and let us know?

**Reply by yogeshnikam671** [reply]
Hey @maheshkhanna, did this solve your issue? Just wanted to follow up on this.
